keps/0009-operator-toolkit.md

```diff
@@ -0,0 +1,296 @@
+
+All tasks within the same step are applied at the same time. For runnable templates within a task (such as a Deployment, StatefulSet, Pod, etc), parallelism of workloads should be controlled through other Kubernetes primitives such as PodDisruptionBudget, and added to the task where the relevant workload is run.
+
+If a step contains a runnable task that has readiness probes defined, the step waits until the readiness probe succeeds before moving onto the next step.
+
+### Tasks
+
+Tasks are a map of task name to a list of templates that should be executed in that step. With [Extensions and Bases](#extensions-and-bases), tasks can be represented in multiple forms, as long as the end result is a ready-to-run YAML manifest for a Kubernetes object.
+
+#### Files
+
+If a filename is specified, KUDO will execute a Go Template on the relevant filename, described more in detail in [Templates](#templates).
+
+#### Webhooks
+
+If a webhook is specified, KUDO will POST the listed webhook with a JSON body containing the full KUDO [execution state](#execution-state). The webhook **MUST** respond with the following:
+
+- Status code: 200
+- Content-type: application/json
+
+The body of this response **MUST** be a fully-qualified Kubernetes object in JSON form. If multiple objects are required, they **MUST** be wrapped in a Kubernetes List API object. The contents of a list object **CAN** be composed of multiple Kubernetes objects of different types.
+
+Webhooks **CAN** have side effects to control external state or to fetch external parameters for use in their templates, as long as the response is still a Kubernetes object.
+
+Webhook HTTP connections time out after 30 seconds, after which a Timeout event is added to the PlanExecution the task is running on.
+
+#### Resources vs. Patches
+
+KUDO additionally supports Kustomize for defining resources and patching for tasks. Kustomize is applied after any Go templating steps. This is useful for defining common bases for objects, or for extending frameworks as described in [Extensions and Bases](#extensions-and-bases).
+
+#### Task Application
+
+Once KUDO has assembled a full set of templates for a task, they will be applied using Kubernetes server-side apply at the same time.
+
+### Parameters
+
+Parameters are a map of Parameter structs, with the key being the name of the parameter. A default value can be set inside of a parameter, as well as a description. When no default is specified, the value is required. If an Instance is created without a required parameter, it will have an Error event added to it describing the validation failure. As described in [Execution State](#execution-state), they are wrapped into the `.Params` object for use by all templated objects. All keys, even arbitrary keys unused by KUDO, are wrapped into the `.Params` object for potential use by other templates.
```

Comment:
so this will be done by applying the instance crd? or how will I actually instantiate the framework? `params.yaml` is just a collection of possible parameters and their defaults, right?

Reply:
If you haven't provided a value, what should the rendering engine use for that value if its not provided by the Application Operator?  Since we don't have any spec for the generation of random values for properties (e.g. database secrets), we need to have everything present at runtime.  Looking through the scenarios of Default/NoDefault and Required/NoRequired, there seems to be a clear association between the two:

**Default Value + Required** - If it's required by the AO, then what does the default value even provide?
**Default Value + Not required** - Since its not required, if the AO doesn't provide it, there is still a valid value to use for rending
**No Default + Not Required** - No value can be used during the rending and we'll get errors
**No Default + Required** - Since there isn't a default, the AO needs to provide it at runtime.

Reply:
@alenkacz This is just the OperatorDeveloper experience, and doesn't touch on the spec of the `Instance` objects (storage) that KUDO leverages internally, i.e. The ApplicationOperator experience.

Reply:
@runyontr I see what you mean. I still was somehow thinking within the old structure.

So if a AO would have a `params.yaml` that has no default specified for one Parameter and we would treat it then as required, what would **required** actually mean in this context? KUDO would stop with an error message if in the `params.yaml` was not set with a value or when it wasn't set in the `Instance` object?